src/frontend/screens/Settings/ProjectSettings/DeviceName/index.tsx

```diff
@@ -0,0 +1,223 @@
+import * as React from 'react';
+import {Alert, ScrollView, StyleSheet} from 'react-native';
+import {NativeStackNavigationOptions} from '@react-navigation/native-stack';
+import {MessageDescriptor, defineMessages, useIntl} from 'react-intl';
+import {useForm} from 'react-hook-form';
+
+import {
+  NativeNavigationComponent,
+  NativeRootNavigationProps,
+} from '../../../../sharedTypes';
+import {
+  useDeviceInfo,
+  useEditDeviceInfo,
+} from '../../../../hooks/server/deviceInfo';
+import {Text} from '../../../../sharedComponents/Text';
+import {HookFormTextInput} from '../../../../sharedComponents/HookFormTextInput';
+import {IconButton} from '../../../../sharedComponents/IconButton';
+import {EditIcon, SaveIcon} from '../../../../sharedComponents/icons';
+import {FieldRow} from './FieldRow';
+import {BLACK} from '../../../../lib/styles';
+
+const m = defineMessages({
+  title: {
+    id: 'screens.Setting.ProjectSettings.DeviceName.title',
+    defaultMessage: 'Device Name',
+  },
+  deviceNameDisplayHeader: {
+    id: 'screens.Setting.ProjectSettings.DeviceName.deviceNameDisplayHeader',
+    defaultMessage: 'Your Device Name',
+  },
+  deviceNameEditHeader: {
+    id: 'screens.Setting.ProjectSettings.DeviceName.deviceNameEditHeader',
+    defaultMessage: 'Edit Device Name',
+  },
+  discardChangesAlertTitle: {
+    id: 'screens.Setting.ProjectSettings.DeviceName.discardChangesAlertTitle',
+    defaultMessage: 'Discard changes?',
+  },
+  discardChangesAlertConfirmText: {
+    id: 'screens.Setting.ProjectSettings.DeviceName.discardChangesAlertConfirmText',
+    defaultMessage: 'Discard Changes',
+  },
+  discardChangesAlertCancelText: {
+    id: 'screens.Setting.ProjectSettings.DeviceName.discardChangesAlertCancelText',
+    defaultMessage: 'Continue Editing',
+  },
+});
+
+export type NavigationParams = {mode: 'display' | 'edit'};
+
+export const DeviceName: NativeNavigationComponent<'DeviceName'> = ({
+  route,
+  navigation,
+}) => {
+  const {mode} = route.params;
+  const {data} = useDeviceInfo();
+
+  return (
+    <ScrollView contentContainerStyle={styles.container}>
+      {data?.name &&
+        (mode === 'display' ? (
+          <DisplayMode deviceName={data.name} navigation={navigation} />
+        ) : (
+          <EditMode deviceName={data.name} navigation={navigation} />
+        ))}
+    </ScrollView>
+  );
+};
+
+function DisplayMode({
+  deviceName,
+  navigation,
+}: {
+  deviceName: string;
+  navigation: React.ComponentProps<
+    NativeNavigationComponent<'DeviceName'>
+  >['navigation'];
+}) {
+  const {formatMessage: t} = useIntl();
+
+  React.useEffect(
+    function updateNavigationOptions() {
+      navigation.setOptions({
+        headerRight: () => (
+          <IconButton onPress={() => navigation.setParams({mode: 'edit'})}>
+            <EditIcon />
+          </IconButton>
+        ),
+      });
+    },
+    [navigation],
+  );
```

Comment:
not a fan of this pattern from React Navigation but it's necessary because we update the nav header bar based on a navigation param value, so it needs to re-render the headers each time we set those params. to my surprise, the `options` field for the Screen component doesn't do that, as it seems to only handle the case of navigating to the screen initiallly